File: src/windows/AmbientContextMcp.Core/Hub/LocalContextHub.cs
using AmbientContextMcp.Core.Models;
using AmbientContextMcp.Core.Settings;

namespace AmbientContextMcp.Core.Hub;

public sealed class LocalContextHub
{
    private const int DefaultMaxEventAgeHours = 24;
    private const int DefaultMaxEventCount = 500;
    private const int MinMaxEventAgeHours = 1;
    private const int MaxMaxEventAgeHours = 168;
    private const int MinMaxEventCount = 100;
    private const int MaxMaxEventCount = 5000;
    private const int DefaultPollLimit = 50;
    private const int MaxPollLimit = 1000;

    private readonly object _lock = new();
    private readonly List<LocalContextEvent> _events = [];
    private readonly HashSet<string> _eventFingerprints = new(StringComparer.Ordinal);
    private readonly LocalContextCursorTracker _cursorTracker = new();
    private readonly ISettingsStore _settingsStore;
    private readonly LocalContextEventLog _eventLog;

    private IReadOnlyList<AmbientState> _latestStates = [];
    private IReadOnlyList<PrivacyClassification> _privacyClassifications = [];
    private AmbientTransmissionPolicySnapshot _transmissionPolicy = new();
    private DateTimeOffset _latestObservedAt;
    private int _observedStateCount;
    private int _outboundStateCount;
    private int _internalEventHistoryCount;
    private int _outboundEventCandidateCount;
    private int _maxEventAgeHours = DefaultMaxEventAgeHours;
    private int _maxEventCount = DefaultMaxEventCount;
    private bool _persistEventLog;
    private long _nextSequence;
    private string _policyVersion = "";

    public event EventHandler<LocalContextEvent>? EventPublished;

    public LocalContextHub(ISettingsStore settingsStore)
    {
        _settingsStore = settingsStore ?? throw new ArgumentNullException(nameof(settingsStore));
        _eventLog = new LocalContextEventLog(LocalContextEventLog.ResolvePath(_settingsStore.SettingsPath));
        ApplySettings(initialLoad: true);
        if (_persistEventLog)
        {
            LoadPersistedEventLog();
        }
    }

    public void Ingest(AmbientContextSnapshot snapshot)
    {
        List<LocalContextEvent> publishedEvents = [];
        var trimmedAny = false;
        lock (_lock)
        {
            _latestObservedAt = snapshot.ObservedAt;
            _latestStates = snapshot.OutboundStates.ToList();
            _privacyClassifications = snapshot.PrivacyClassifications.ToList();
            _transmissionPolicy = snapshot.TransmissionPolicy;
            _policyVersion = PolicyVersionService.ComputePolicyVersion(
                _privacyClassifications,
                _transmissionPolicy.PathTransmitOverrides);

            // events.jsonl から復元した旧スキーマのイベントは PayloadSensitivity が空のまま
            // _events に乗っているため、scope フィルタが event-level だけにフォールバックして
            // 高機微 payload キーが意図せず流れ続ける。最初に classifications を持つ snapshot が
            // 来たタイミングで一度だけ backfill する (以降は一致するエントリが無く no-op)。
            var backfilled = BackfillLegacyPayloadSensitivity();
            _observedStateCount = snapshot.States.Count;
            _outboundStateCount = snapshot.OutboundStates.Count;
            _internalEventHistoryCount = snapshot.Events.Count;
            _outboundEventCandidateCount = snapshot.OutboundEvents.Count;

            foreach (var outboundEvent in snapshot.OutboundEvents)
            {
                var fingerprint = GetFingerprint(outboundEvent);
                if (!_eventFingerprints.Add(fingerprint))
                {
                    continue;
                }

                var sequence = ++_nextSequence;
                var (payloadSensitivity, maxFieldSensitivity) = SensitivityScopeFilter.ComputePayloadSensitivity(
                    outboundEvent.Name,
                    outboundEvent.Payload,
                    _privacyClassifications,
                    outboundEvent.Sensitivity);
                var localEvent = new LocalContextEvent
                {
                    Id = CreateEventId(outboundEvent.ObservedAt, sequence),
                    Sequence = sequence,
                    ObservedAt = outboundEvent.ObservedAt,
                    Name = outboundEvent.Name,
                    Value = outboundEvent.Value,
                    Payload = outboundEvent.Payload,
                    Sensitivity = outboundEvent.Sensitivity,
                    PayloadSensitivity = payloadSensitivity,
                    MaxFieldSensitivity = maxFieldSensitivity
                };
                _events.Add(localEvent);
                publishedEvents.Add(localEvent);
            }

            trimmedAny = TrimEvents(DateTimeOffset.Now);

            if (_persistEventLog)
            {
                if (trimmedAny || backfilled)
                {
                    // 古いイベントが落ちたか、または旧スキーマの backfill が走ったので
                    // JSONL を _events から書き直す (compaction)。
                    // 同一トランザクションの新規イベントもこの 1 回の書き出しに含まれる。
                    RewriteEventLogUnlocked();
                }
                else if (publishedEvents.Count > 0)
                {
                    AppendEventsUnlocked(publishedEvents);
                }
            }
        }

        foreach (var localEvent in publishedEvents)
        {
            EventPublished?.Invoke(this, localEvent);
        }
    }

    public LocalContextStateResponse GetContextStates(LocalContextStateRequest request)
    {
        lock (_lock)
        {
            var states = _latestStates
                .Where(state => IsNameIncluded(state.Name, request.Names))
                .Where(state => SensitivityScopeFilter.IsSensitivityAllowed(state.Sensitivity, request.Scopes))
                .Select(state => ToLocalContextState(state, request.IncludeMetadata))
                .ToList();

            return new LocalContextStateResponse
            {
                ObservedAt = _latestObservedAt,
                States = states,
                PolicyVersion = _policyVersion
            };
        }
    }

    public LocalContextPollResponse PollEvents(LocalContextPollRequest request)
    {
        lock (_lock)
        {
            TrimEvents(DateTimeOffset.Now);

            var limit = NormalizeLimit(request.Limit);
            var isHistoryQuery = request.Since.HasValue || request.Until.HasValue;
            var cursorResult = _cursorTracker.Resolve(
                request.ClientId,
                request.Cursor,
                isHistoryQuery,
                FirstSequence(),
                LatestSequence());

            var matchingEvents = _events
                .Where(item => item.Sequence > cursorResult.Sequence)
                .Where(item => !request.Since.HasValue || item.ObservedAt >= request.Since.Value)
                .Where(item => !request.Until.HasValue || item.ObservedAt <= request.Until.Value)
                .Where(item => IsNameIncluded(item.Name, request.Names))
                .Select(item => SensitivityScopeFilter.FilterEventForScope(item, request.Scopes))
                .Where(item => item is not null)
                .Select(item => request.IncludePayload ? item! : StripPayload(item!))
                .Take(limit + 1)
                .ToList();

            var returnedEvents = matchingEvents.Take(limit).ToList();
            var lastSequence = returnedEvents.Count > 0
                ? returnedEvents[^1].Sequence
                : cursorResult.Sequence;

            // history query (since/until 指定) は副作用なしの stateless 取得。
            // クライアント位置を進めない (= 同じ範囲で再取得しても結果が消えない)。
            // pagination は呼び出し側が NextCursor を渡すことで行う。
            if (!isHistoryQuery)
            {
                _cursorTracker.Advance(request.ClientId, lastSequence);
            }

            return new LocalContextPollResponse
            {
                Events = returnedEvents,
                NextCursor = LocalContextCursorTracker.Encode(lastSequence),
                HasMore = matchingEvents.Count > limit,
                CursorExpired = cursorResult.Expired,
                Retention = new LocalContextRetentionInfo
                {
                    MaxAgeHours = _maxEventAgeHours,
                    MaxEvents = _maxEventCount
                },
                PolicyVersion = _policyVersion
            };
        }
    }

    public LocalContextEventSchemasResponse GetEventSchemas()
    {
        return new LocalContextEventSchemasResponse
        {
            Events = EventSchemaCatalog.GetAll()
        };
    }

    public LocalContextPolicyResponse GetPolicy()
    {
        lock (_lock)
        {
            return new LocalContextPolicyResponse
            {
                ObservedAt = _latestObservedAt,
                TransmissionPolicy = _transmissionPolicy,
                PrivacyClassifications = _privacyClassifications.ToList(),
                EffectivePolicies = _privacyClassifications
                    .Select(item => BuildEffectivePolicy(item, _transmissionPolicy))
                    .ToList(),
                ObservedStateCount = _observedStateCount,
                OutboundStateCount = _outboundStateCount,
                InternalEventHistoryCount = _internalEventHistoryCount,
                OutboundEventCandidateCount = _outboundEventCandidateCount,
                RetainedOutboundEventCount = _events.Count,
                Retention = new LocalContextRetentionInfo
                {
                    MaxAgeHours = _maxEventAgeHours,
                    MaxEvents = _maxEventCount
                }
            };
        }
    }

    public void ReloadSettings()
    {
        ApplySettings(initialLoad: false);
    }

    /// <summary>
    /// 設定の読み込みと、永続化フラグ遷移に応じたファイル操作を一箇所にまとめる。
    /// initialLoad = true (= コンストラクタからの初回呼び出し) の場合、
    /// 「OFF→ON / ON→OFF 同期」分岐を**意図的にスキップ**する。
    /// 初回時点で _persistEventLog は default false で初期化されているため、
    /// ユーザーが永続化を ON にした保存設定をロードすると false→true に見えてしまうが、
    /// それを「ユーザーが ON に切替えた」と誤認して空の _events をファイルに書き戻すと、
    /// 既存の events.jsonl を消してしまう (これが過去のバグ要因)。
    /// 初回の events.jsonl 復元はコンストラクタ側の <see cref="LoadPersistedEventLog"/> に任せる。
    /// </summary>
    private void ApplySettings(bool initialLoad)
    {
        var settings = _settingsStore.LoadLocalContextSettings();
        lock (_lock)
        {
            var previousPersist = _persistEventLog;
            _maxEventAgeHours = NormalizeMaxEventAgeHours(settings.MaxEventAgeHours);
            _maxEventCount = NormalizeMaxEventCount(settings.MaxEventCount);
            _persistEventLog = settings.PersistEventLog;
            var trimmed = TrimEvents(DateTimeOffset.Now);

            if (initialLoad)
            {
                return;
            }

            if (_persistEventLog)
            {
                if (!previousPersist || trimmed)
                {
                    // OFF → ON: 在席中の opt-in。in-memory の _events をファイルに同期する。
                    // trimmed: 古いイベントが落ちたので compaction する。
                    RewriteEventLogUnlocked();
                }
            }
            else if (previousPersist)
            {
                // ON → OFF: ユーザーが明示的に opt-out したのでファイルを削除する。
                DeleteEventLogUnlocked();
            }
        }
    }

    /// <summary>
    /// 戻り値は「実際にイベントを 1 件以上落としたか」。永続化が ON の場合、
    /// 呼び出し側はこのフラグを見て events.jsonl を rewrite するか判定する。
    /// </summary>
    /// <summary>
    /// PayloadSensitivity / MaxFieldSensitivity が空のままの _events エントリを
    /// 現在の _privacyClassifications から再計算して詰め直す。
    /// アップグレード直後の events.jsonl 復元エントリ (= 旧スキーマ) を新フィルタが正しく
    /// 間引けるようにする目的。返り値は実際に 1 件以上書き換えたかどうか。
    /// </summary>
    private bool BackfillLegacyPayloadSensitivity()
    {
        var backfilled = false;
        for (var i = 0; i < _events.Count; i++)
        {
            var current = _events[i];
            if (!string.IsNullOrEmpty(current.MaxFieldSensitivity))
            {
                continue;
            }

            var (perKey, max) = SensitivityScopeFilter.ComputePayloadSensitivity(
                current.Name,
                current.Payload,
                _privacyClassifications,
                current.Sensitivity);
            _events[i] = new LocalContextEvent
            {
                Id = current.Id,
                Sequence = current.Sequence,
                ObservedAt = current.ObservedAt,
                Name = current.Name,
                Value = current.Value,
                Payload = current.Payload,
                Sensitivity = current.Sensitivity,
                PayloadSensitivity = perKey,
                MaxFieldSensitivity = max
            };
            backfilled = true;
        }

        return backfilled;
    }

    private bool TrimEvents(DateTimeOffset now)
    {
        var trimmed = false;
        var cutoff = now - TimeSpan.FromHours(_maxEventAgeHours);
        while (_events.Count > 0 && _events[0].ObservedAt < cutoff)
        {
            _eventFingerprints.Remove(GetFingerprint(_events[0]));
            _events.RemoveAt(0);
            trimmed = true;
        }

        while (_events.Count > _maxEventCount)
        {
            _eventFingerprints.Remove(GetFingerprint(_events[0]));
            _events.RemoveAt(0);
            trimmed = true;
        }

        return trimmed;
    }

    /// <summary>
    /// 起動時に既存 events.jsonl を読み込んで _events / _eventFingerprints / _nextSequence を再構築する。
    /// 読み込み後、現在の age/count 制限で trim する。
    /// </summary>
    private void LoadPersistedEventLog()
    {
        lock (_lock)
        {
            foreach (var loaded in _eventLog.Load())
            {
                var fingerprint = GetFingerprint(loaded);
                if (!_eventFingerprints.Add(fingerprint))
                {
                    continue;
                }

                _events.Add(loaded);
                if (loaded.Sequence > _nextSequence)
                {
                    _nextSequence = loaded.Sequence;
                }
            }

            _events.Sort((a, b) => a.Sequence.CompareTo(b.Sequence));

            if (TrimEvents(DateTimeOffset.Now))
            {
                RewriteEventLogUnlocked();
            }
        }
    }

    private void AppendEventsUnlocked(IReadOnlyList<LocalContextEvent> events)
    {
        _eventLog.Append(events);
    }

    private void RewriteEventLogUnlocked()
    {
        _eventLog.Rewrite(_events);
    }

    private void DeleteEventLogUnlocked()
    {
        _eventLog.Delete();
    }

    public static int NormalizeMaxEventAgeHours(int hours)
    {
        if (hours <= 0)
        {
            return DefaultMaxEventAgeHours;
        }

        return Math.Clamp(hours, MinMaxEventAgeHours, MaxMaxEventAgeHours);
    }

    public static int NormalizeMaxEventCount(int count)
    {
        if (count <= 0)
        {
            return DefaultMaxEventCount;
        }

        return Math.Clamp(count, MinMaxEventCount, MaxMaxEventCount);
    }

    private long FirstSequence()
    {
        return _events.Count == 0 ? 0 : _events[0].Sequence;
    }

    private long LatestSequence()
    {
        return _events.Count == 0 ? 0 : _events[^1].Sequence;
    }

    private static int NormalizeLimit(int limit)
    {
        if (limit <= 0)
        {
            return DefaultPollLimit;
        }

        return Math.Min(limit, MaxPollLimit);
    }

    private static bool IsNameIncluded(string name, IReadOnlyList<string> names)
    {
        return names.Count == 0 || names.Contains(name, StringComparer.OrdinalIgnoreCase);
    }

    private static LocalContextState ToLocalContextState(AmbientState state, bool includeMetadata)
    {
        return new LocalContextState
        {
            ObservedAt = includeMetadata ? state.ObservedAt : null,
            Name = state.Name,
            Value = state.Value,
            Sensitivity = includeMetadata ? state.Sensitivity : null
        };
    }

    private static LocalContextEffectivePolicy BuildEffectivePolicy(
        PrivacyClassification classification,
        AmbientTransmissionPolicySnapshot transmissionPolicy)
    {
        var hasOverride = TryGetOverride(
            classification.Path,
            transmissionPolicy.PathTransmitOverrides,
            out var overridePath,
            out var overrideTransmit);

        return new LocalContextEffectivePolicy
        {
            Path = classification.Path,
            Sensitivity = classification.Sensitivity,
            RequiredScope = "context." + SensitivityScopeFilter.NormalizeSensitivity(classification.Sensitivity) + ":read",
            DefaultTransmit = classification.DefaultTransmit,
            EffectiveTransmit = hasOverride ? overrideTransmit : classification.DefaultTransmit,
            HasOverride = hasOverride,
            OverridePath = overridePath,
            OverrideTransmit = hasOverride ? overrideTransmit : null,
            Reason = classification.Reason
        };
    }

    private static bool TryGetOverride(
        string path,
        IReadOnlyDictionary<string, bool> overrides,
        out string overridePath,
        out bool allowed)
    {
        var current = path;
        while (!string.IsNullOrWhiteSpace(current))
        {
            if (overrides.TryGetValue(current, out allowed))
            {
                overridePath = current;
                return true;
            }

            var lastDot = current.LastIndexOf('.');
            if (lastDot < 0)
            {
                break;
            }

            current = current[..lastDot];
        }

        overridePath = "";
        allowed = false;
        return false;
    }

    /// <summary>
    /// 要約モードで返すために payload と payloadSensitivity を空にした複製を作る。
    /// id / sequence / observedAt / name / value / sensitivity / maxFieldSensitivity は保持する。
    /// </summary>
    private static LocalContextEvent StripPayload(LocalContextEvent ev)
    {
        return new LocalContextEvent
        {
            Id = ev.Id,
            Sequence = ev.Sequence,
            ObservedAt = ev.ObservedAt,
            Name = ev.Name,
            Value = ev.Value,
            Sensitivity = ev.Sensitivity,
            MaxFieldSensitivity = ev.MaxFieldSensitivity
            // Payload / PayloadSensitivity は init 既定の空 dict のまま
        };
    }

    private static string CreateEventId(DateTimeOffset observedAt, long sequence)
    {
        return $"evt_{observedAt.UtcDateTime:yyyyMMddHHmmssfff}_{sequence:D6}";
    }

    private static string GetFingerprint(AmbientOutboundEvent outboundEvent)
    {
        return string.Join(
            "|",
            outboundEvent.ObservedAt.ToUnixTimeMilliseconds().ToString(System.Globalization.CultureInfo.InvariantCulture),
            outboundEvent.Name,
            outboundEvent.Value,
            GetPayloadFingerprint(outboundEvent.Payload));
    }

    private static string GetFingerprint(LocalContextEvent localEvent)
    {
        return string.Join(
            "|",
            localEvent.ObservedAt.ToUnixTimeMilliseconds().ToString(System.Globalization.CultureInfo.InvariantCulture),
            localEvent.Name,
            localEvent.Value,
            GetPayloadFingerprint(localEvent.Payload));
    }

    private static string GetPayloadFingerprint(IReadOnlyDictionary<string, string> payload)
    {
        return string.Join(
            "&",
            payload
                .OrderBy(item => item.Key, StringComparer.OrdinalIgnoreCase)
                .Select(item => item.Key + "=" + item.Value));
    }
}

File: src/macos/Sources/AmbientContextCore/Hub/LocalContextHub.swift
import Foundation

/// C# 版 `LocalContextHub` の移植。
/// C# は `lock (_lock)` で保護していたので、Swift では `NSLock` を持つ `final class` にし、
/// `@unchecked Sendable` を付けて公開 API (ingest / getContextStates / pollEvents /
/// getEventSchemas / getPolicy / reloadSettings / eventPublished) をそのまま保つ。
public final class LocalContextHub: @unchecked Sendable {
    private static let defaultMaxEventAgeHours = 24
    private static let defaultMaxEventCount = 500
    private static let minMaxEventAgeHours = 1
    private static let maxMaxEventAgeHours = 168
    private static let minMaxEventCount = 100
    private static let maxMaxEventCount = 5000
    private static let defaultPollLimit = 50
    private static let maxPollLimit = 1000

    private let lock = NSLock()
    private var events: [LocalContextEvent] = []
    private var eventFingerprints = Set<String>()
    private let cursorTracker = LocalContextCursorTracker()
    private let settingsStore: any SettingsStore
    private let eventLog: LocalContextEventLog
    private let language: String

    private var latestStates: [AmbientState] = []
    private var privacyClassifications: [PrivacyClassification] = []
    private var transmissionPolicy = AmbientTransmissionPolicySnapshot()
    private var latestObservedAt = Date(timeIntervalSince1970: 0)
    private var observedStateCount = 0
    private var outboundStateCount = 0
    private var internalEventHistoryCount = 0
    private var outboundEventCandidateCount = 0
    private var maxEventAgeHours = LocalContextHub.defaultMaxEventAgeHours
    private var maxEventCount = LocalContextHub.defaultMaxEventCount
    private var persistEventLog = false
    private var nextSequence: Int64 = 0
    private var policyVersion = ""

    /// C# の `event EventHandler<LocalContextEvent> EventPublished` 相当。
    public var eventPublished: (@Sendable (LocalContextEvent) -> Void)?

    public init(settingsStore: any SettingsStore, language: String = AmbientContextCatalog.defaultLanguage) {
        self.settingsStore = settingsStore
        self.language = language
        self.eventLog = LocalContextEventLog(
            path: LocalContextEventLog.resolvePath(settingsPath: settingsStore.settingsPath))
        applySettings(initialLoad: true)
        if persistEventLog {
            loadPersistedEventLog()
        }
    }

    public func ingest(_ snapshot: AmbientContextSnapshot) {
        var publishedEvents: [LocalContextEvent] = []
        lock.lock()

        latestObservedAt = snapshot.observedAt
        latestStates = snapshot.outboundStates
        privacyClassifications = snapshot.privacyClassifications
        transmissionPolicy = snapshot.transmissionPolicy
        policyVersion = PolicyVersionService.computePolicyVersion(
            classifications: privacyClassifications,
            overrides: transmissionPolicy.pathTransmitOverrides)

        // events.jsonl から復元した旧スキーマのイベントは payloadSensitivity が空のまま
        // events に乗っているため、scope フィルタが event-level だけにフォールバックして
        // 高機微 payload キーが意図せず流れ続ける。最初に classifications を持つ snapshot が
        // 来たタイミングで一度だけ backfill する (以降は一致するエントリが無く no-op)。
        let backfilled = backfillLegacyPayloadSensitivity()
        observedStateCount = snapshot.states.count
        outboundStateCount = snapshot.outboundStates.count
        internalEventHistoryCount = snapshot.events.count
        outboundEventCandidateCount = snapshot.outboundEvents.count

        for outboundEvent in snapshot.outboundEvents {
            let fingerprint = Self.fingerprint(
                observedAt: outboundEvent.observedAt,
                name: outboundEvent.name,
                value: outboundEvent.value,
                payload: outboundEvent.payload)
            if eventFingerprints.contains(fingerprint) { continue }
            eventFingerprints.insert(fingerprint)

            nextSequence += 1
            let sequence = nextSequence
            let computed = SensitivityScopeFilter.computePayloadSensitivity(
                eventName: outboundEvent.name,
                payload: outboundEvent.payload,
                classifications: privacyClassifications,
                eventSensitivity: outboundEvent.sensitivity)
            let localEvent = LocalContextEvent(
                id: Self.createEventId(observedAt: outboundEvent.observedAt, sequence: sequence),
                sequence: sequence,
                observedAt: outboundEvent.observedAt,
                name: outboundEvent.name,
                value: outboundEvent.value,
                payload: outboundEvent.payload,
                sensitivity: outboundEvent.sensitivity,
                payloadSensitivity: computed.perKey,
                maxFieldSensitivity: computed.max)
            events.append(localEvent)
            publishedEvents.append(localEvent)
        }

        let trimmedAny = trimEvents(now: Date())

        if persistEventLog {
            if trimmedAny || backfilled {
                // 古いイベントが落ちたか、または旧スキーマの backfill が走ったので
                // JSONL を events から書き直す (compaction)。
                // 同一トランザクションの新規イベントもこの 1 回の書き出しに含まれる。
                eventLog.rewrite(events)
            } else if !publishedEvents.isEmpty {
                eventLog.append(publishedEvents)
            }
        }

        lock.unlock()

        for localEvent in publishedEvents {
            eventPublished?(localEvent)
        }
    }

    public func getContextStates(_ request: LocalContextStateRequest) -> LocalContextStateResponse {
        lock.lock()
        defer { lock.unlock() }

        let states = latestStates
            .filter { Self.isNameIncluded($0.name, request.names) }
            .filter { SensitivityScopeFilter.isSensitivityAllowed($0.sensitivity, scopes: request.scopes) }
            .map { Self.toLocalContextState($0, includeMetadata: request.includeMetadata) }

        return LocalContextStateResponse(
            observedAt: latestObservedAt,
            states: states,
            policyVersion: policyVersion)
    }

    public func pollEvents(_ request: LocalContextPollRequest) -> LocalContextPollResponse {
        lock.lock()
        defer { lock.unlock() }

        let now = Date()
        _ = trimEvents(now: now)
        cursorTracker.pruneStale(now: now)

        let limit = Self.normalizeLimit(request.limit)
        let isHistoryQuery = request.since != nil || request.until != nil
        let cursorResult = cursorTracker.resolve(
            clientId: request.clientId,
            cursor: request.cursor,
            isHistoryQuery: isHistoryQuery,
            firstSequence: firstSequence(),
            latestSequence: latestSequence(),
            now: now)

        var matchingEvents: [LocalContextEvent] = []
        for item in events {
            guard item.sequence > cursorResult.sequence else { continue }
            if let since = request.since, item.observedAt < since { continue }
            if let until = request.until, item.observedAt > until { continue }
            guard Self.isNameIncluded(item.name, request.names) else { continue }
            guard let filtered = SensitivityScopeFilter.filterEventForScope(item, scopes: request.scopes) else {
                continue
            }
            matchingEvents.append(request.includePayload ? filtered : Self.stripPayload(filtered))
            if matchingEvents.count > limit { break }
        }

        let returnedEvents = Array(matchingEvents.prefix(limit))
        let lastSequence = returnedEvents.last?.sequence ?? cursorResult.sequence

        // history query (since/until 指定) は副作用なしの stateless 取得。
        // クライアント位置を進めない (= 同じ範囲で再取得しても結果が消えない)。
        // pagination は呼び出し側が nextCursor を渡すことで行う。
        if !isHistoryQuery {
            cursorTracker.advance(clientId: request.clientId, sequence: lastSequence, now: now)
        }

        return LocalContextPollResponse(
            events: returnedEvents,
            nextCursor: LocalContextCursorTracker.encode(lastSequence),
            hasMore: matchingEvents.count > limit,
            cursorExpired: cursorResult.expired,
            retention: LocalContextRetentionInfo(
                maxAgeHours: maxEventAgeHours,
                maxEvents: maxEventCount),
            policyVersion: policyVersion)
    }

    public func getEventSchemas() -> LocalContextEventSchemasResponse {
        LocalContextEventSchemasResponse(
            events: AmbientContextCatalog.getEventSchemas(language: language))
    }

    public func getPolicy() -> LocalContextPolicyResponse {
        lock.lock()
        defer { lock.unlock() }

        return LocalContextPolicyResponse(
            observedAt: latestObservedAt,
            transmissionPolicy: transmissionPolicy,
            privacyClassifications: privacyClassifications,
            effectivePolicies: privacyClassifications.map {
                Self.buildEffectivePolicy($0, transmissionPolicy: transmissionPolicy)
            },
            observedStateCount: observedStateCount,
            outboundStateCount: outboundStateCount,
            internalEventHistoryCount: internalEventHistoryCount,
            outboundEventCandidateCount: outboundEventCandidateCount,
            retainedOutboundEventCount: events.count,
            retention: LocalContextRetentionInfo(
                maxAgeHours: maxEventAgeHours,
                maxEvents: maxEventCount))
    }

    public func reloadSettings() {
        applySettings(initialLoad: false)
    }

    /// 設定の読み込みと、永続化フラグ遷移に応じたファイル操作を一箇所にまとめる。
    /// initialLoad = true (= イニシャライザからの初回呼び出し) の場合、
    /// 「OFF→ON / ON→OFF 同期」分岐を**意図的にスキップ**する。
    /// 初回時点で persistEventLog は false で初期化されているため、ユーザーが永続化を ON にした
    /// 保存設定をロードすると false→true に見えてしまうが、それを「ユーザーが ON に切替えた」と
    /// 誤認して空の events をファイルに書き戻すと、既存の events.jsonl を消してしまう。
    /// 初回の events.jsonl 復元は `loadPersistedEventLog` に任せる。
    private func applySettings(initialLoad: Bool) {
        let settings = settingsStore.loadLocalContextSettings()
        lock.lock()
        defer { lock.unlock() }

        let previousPersist = persistEventLog
        maxEventAgeHours = Self.normalizeMaxEventAgeHours(settings.maxEventAgeHours)
        maxEventCount = Self.normalizeMaxEventCount(settings.maxEventCount)
        persistEventLog = settings.persistEventLog
        let trimmed = trimEvents(now: Date())

        if initialLoad {
            return
        }

        if persistEventLog {
            if !previousPersist || trimmed {
                // OFF → ON: 在席中の opt-in。in-memory の events をファイルに同期する。
                // trimmed: 古いイベントが落ちたので compaction する。
                eventLog.rewrite(events)
            }
        } else if previousPersist {
            // ON → OFF: ユーザーが明示的に opt-out したのでファイルを削除する。
            eventLog.delete()
        }
    }

    /// payloadSensitivity / maxFieldSensitivity が空のままの events エントリを
    /// 現在の privacyClassifications から再計算して詰め直す。返り値は 1 件以上書き換えたかどうか。
    private func backfillLegacyPayloadSensitivity() -> Bool {
        var backfilled = false
        for index in events.indices {
            let current = events[index]
            if !current.maxFieldSensitivity.isEmpty { continue }

            let computed = SensitivityScopeFilter.computePayloadSensitivity(
                eventName: current.name,
                payload: current.payload,
                classifications: privacyClassifications,
                eventSensitivity: current.sensitivity)
            events[index] = LocalContextEvent(
                id: current.id,
                sequence: current.sequence,
                observedAt: current.observedAt,
                name: current.name,
                value: current.value,
                payload: current.payload,
                sensitivity: current.sensitivity,
                payloadSensitivity: computed.perKey,
                maxFieldSensitivity: computed.max)
            backfilled = true
        }
        return backfilled
    }

    /// 戻り値は「実際にイベントを 1 件以上落としたか」。永続化が ON の場合、
    /// 呼び出し側はこのフラグを見て events.jsonl を rewrite するか判定する。
    private func trimEvents(now: Date) -> Bool {
        var trimmed = false
        let cutoff = now.addingTimeInterval(-Double(maxEventAgeHours) * 3600)
        while let first = events.first, first.observedAt < cutoff {
            eventFingerprints.remove(Self.fingerprint(
                observedAt: first.observedAt,
                name: first.name,
                value: first.value,
                payload: first.payload))
            events.removeFirst()
            trimmed = true
        }

        while events.count > maxEventCount {
            let first = events[0]
            eventFingerprints.remove(Self.fingerprint(
                observedAt: first.observedAt,
                name: first.name,
                value: first.value,
                payload: first.payload))
            events.removeFirst()
            trimmed = true
        }

        return trimmed
    }

    /// 起動時に既存 events.jsonl を読み込んで events / eventFingerprints / nextSequence を再構築する。
    /// 読み込み後、現在の age/count 制限で trim する。
    private func loadPersistedEventLog() {
        lock.lock()
        defer { lock.unlock() }

        for loaded in eventLog.load() {
            let fingerprint = Self.fingerprint(
                observedAt: loaded.observedAt,
                name: loaded.name,
                value: loaded.value,
                payload: loaded.payload)
            if eventFingerprints.contains(fingerprint) { continue }
            eventFingerprints.insert(fingerprint)

            events.append(loaded)
            if loaded.sequence > nextSequence {
                nextSequence = loaded.sequence
            }
        }

        events.sort { $0.sequence < $1.sequence }

        if trimEvents(now: Date()) {
            eventLog.rewrite(events)
        }
    }

    public static func normalizeMaxEventAgeHours(_ hours: Int) -> Int {
        if hours <= 0 { return defaultMaxEventAgeHours }
        return min(max(hours, minMaxEventAgeHours), maxMaxEventAgeHours)
    }

    public static func normalizeMaxEventCount(_ count: Int) -> Int {
        if count <= 0 { return defaultMaxEventCount }
        return min(max(count, minMaxEventCount), maxMaxEventCount)
    }

    private func firstSequence() -> Int64 {
        events.first?.sequence ?? 0
    }

    private func latestSequence() -> Int64 {
        events.last?.sequence ?? 0
    }

    private static func normalizeLimit(_ limit: Int) -> Int {
        if limit <= 0 { return defaultPollLimit }
        return min(limit, maxPollLimit)
    }

    private static func isNameIncluded(_ name: String, _ names: [String]) -> Bool {
        names.isEmpty || names.contains { $0.lowercased() == name.lowercased() }
    }

    private static func toLocalContextState(_ state: AmbientState, includeMetadata: Bool) -> LocalContextState {
        LocalContextState(
            observedAt: includeMetadata ? state.observedAt : nil,
            name: state.name,
            value: state.value,
            sensitivity: includeMetadata ? state.sensitivity : nil)
    }

    static func buildEffectivePolicy(
        _ classification: PrivacyClassification,
        transmissionPolicy: AmbientTransmissionPolicySnapshot
    ) -> LocalContextEffectivePolicy {
        let resolved = tryGetOverride(
            path: classification.path,
            overrides: transmissionPolicy.pathTransmitOverrides)

        return LocalContextEffectivePolicy(
            path: classification.path,
            sensitivity: classification.sensitivity,
            requiredScope: "context." +
                SensitivityScopeFilter.normalizeSensitivity(classification.sensitivity) + ":read",
            defaultTransmit: classification.defaultTransmit,
            effectiveTransmit: resolved?.allowed ?? classification.defaultTransmit,
            hasOverride: resolved != nil,
            overridePath: resolved?.path ?? "",
            overrideTransmit: resolved?.allowed,
            reason: classification.reason)
    }

    /// path から `.` 区切りで親方向に遡り、最初に見つかった override を返す。
    static func tryGetOverride(
        path: String,
        overrides: CaseInsensitiveDictionary<Bool>
    ) -> (path: String, allowed: Bool)? {
        var current = path
        while !current.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            if let allowed = overrides[current] {
                return (current, allowed)
            }
            guard let lastDot = current.lastIndex(of: ".") else { break }
            current = String(current[current.startIndex..<lastDot])
        }
        return nil
    }

    /// 要約モードで返すために payload と payloadSensitivity を空にした複製を作る。
    /// id / sequence / observedAt / name / value / sensitivity / maxFieldSensitivity は保持する。
    private static func stripPayload(_ event: LocalContextEvent) -> LocalContextEvent {
        LocalContextEvent(
            id: event.id,
            sequence: event.sequence,
            observedAt: event.observedAt,
            name: event.name,
            value: event.value,
            sensitivity: event.sensitivity,
            maxFieldSensitivity: event.maxFieldSensitivity)
    }

    static func createEventId(observedAt: Date, sequence: Int64) -> String {
        "evt_\(AmbientDateFormat.compactUtcStamp(from: observedAt))_"
            + String(format: "%06lld", sequence)
    }

    private static func fingerprint(
        observedAt: Date,
        name: String,
        value: String,
        payload: CaseInsensitiveDictionary<String>
    ) -> String {
        let unixMilliseconds = Int64((observedAt.timeIntervalSince1970 * 1000).rounded(.down))
        let payloadFingerprint = payload.sortedPairs
            .map { "\($0.key)=\($0.value)" }
            .joined(separator: "&")
        return [String(unixMilliseconds), name, value, payloadFingerprint].joined(separator: "|")
    }
}
